CRM/Event/BAO/Participant.php

```diff
@@ -832,6 +844,324 @@ public static function deleteParticipant($id) {
+      $eventId,
+      TRUE,
+      TRUE,
+      $excludeStatus,
+      $participantRoleIds
+    );
+    $errors = [];
+
+    if (count($existingParticipants) > 0) {
+      if ($isAdditional) {
+        $errors[] = ts("It looks like this participant is already registered for this event. If you want to change your registration, or you feel that you've received this message in error, please contact the site administrator.");
+      }
+      elseif (!(bool) $event['allow_same_participant_emails']) {
+        if ('admin' === $context) {
+          $errors['event_id'] = ts('This contact has already been assigned to this event.');
+        }
+        else {
+          if ('On waitlist' === reset($existingParticipants)['status_id:name']) {
+            $errors[] = ts("It looks like you are already waitlisted for this event. If you want to change your registration, or you feel that you've received this message in error, please contact the site administrator.");
+          }
+          else {
+            $errors[] = ts("It looks like you are already registered for this event. If you want to change your registration, or you feel that you've received this message in error, please contact the site administrator.");
+          }
+        }
+      }
+    }
+
+    return $errors;
+  }
+
+  /**
+   * @param array $values
+   *   A list of values to validate, keyed by participant properties/field names.
+   * @param string $context
+   *   The context for which to create error messages. Either "public" or "admin".
+   *
+   * @return array<int|string,string>
+   *   A list of validation error messages, possibly keyed by affected participant properties/field names.
+   *
+   * @internal
+   */
+  public static function validateAvailableSpaces(
+    array $values,
+    string $context = 'public'
+  ): array {
+    $errors = [];
+    $eventId = $values['event_id'];
+    $event = \Civi\Api4\Event::get(FALSE)
+      ->addSelect('has_waitlist', 'max_participants')
+      ->addWhere('id', '=', $eventId)
+      ->execute()
+      ->single();
+
+    $availableSpaces = CRM_Event_BAO_Participant::eventFull(
+      $eventId,
+      TRUE,
+      $event['has_waitlist']
+    );
+    $spacesAvailable = is_numeric($availableSpaces) ? (int) $availableSpaces : 0;
+
+    //check for availability of registrations.
+    if ($event['max_participants'] !== NULL
+      && !empty($values['additional_participants'])
+      && empty($values['bypass_payment']) &&
+      ((int) $values['additional_participants']) >= $spacesAvailable
+    ) {
+      $errors['additional_participants'] = ts(
+        'There is only enough space left on this event for %1 participant(s).',
+        [1 => $spacesAvailable]
+      );
+    }
+
+    return $errors;
+  }
+
+  /**
+   * @param int $eventId
+   *
+   * @return array<int|string,string>
+   *   A list of validation error messages, possibly keyed by affected participant properties/field names.
+   * @throws \CRM_Core_Exception
+   *
+   * @internal
```

Comment:
I'd really like to go further with these messages and explictly say that the functions are expected to change and cannot be relied on by external code. I worry that the comments are that the budget are running out & we could go a while between this preliminary refactor & having a supported API and people might feel tempted to call them directly so I want the 'at own risk, absolutely not supported until there is an api, may change often' to be as clear as possible